src/control/lib/hostlist/hostlist.go

```diff
@@ -0,0 +1,735 @@
+
+	for _, hr := range other.ranges {
+		if err := hl.pushRange(hr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func fmtRangeHost(hr *hostRange, num uint) string {
+	return fmt.Sprintf("%s%0*d%s", hr.prefix, hr.width, num, hr.suffix)
+}
+
+// Pop returns the string representation of the last host pushed
+// onto the HostList and then removes it from the HostList. Returns
+// an error if the HostList is empty.
+func (hl *HostList) Pop() (hostName string, err error) {
+	if hl.IsEmpty() {
+		return "", ErrEmpty
+	}
+
+	hl.Lock()
+	defer hl.Unlock()
+
+	hl.hostCount--
+
+	tailIdx := len(hl.ranges) - 1
+	tail := hl.ranges[tailIdx]
+	if !tail.isRange {
+		hostName = tail.prefix
+		hl.ranges = hl.ranges[:tailIdx]
+		return
+	}
+
+	hostName = fmtRangeHost(tail, tail.hi)
+	tail.hi--
+	if tail.hi < tail.lo {
+		hl.ranges = hl.ranges[:tailIdx]
+	}
+
+	return
+}
+
+// Shift returns the string representation of the first host pushed
+// onto the HostList and then removes it from the HostList. Returns
+// an error if the HostList is empty.
+func (hl *HostList) Shift() (hostName string, err error) {
+	if hl.IsEmpty() {
+		return "", ErrEmpty
+	}
+
+	hl.Lock()
+	defer hl.Unlock()
+
+	hl.hostCount--
+
+	head := hl.ranges[0]
+	if !head.isRange {
+		hostName = head.prefix
+		hl.ranges = hl.ranges[1:]
+		return
+	}
+
+	hostName = fmtRangeHost(head, head.lo)
+	head.lo++
+	if head.lo > head.hi {
+		hl.ranges = hl.ranges[1:]
+	}
+
+	return
+}
+
+func (hl *HostList) getNthHostRange(n int) (int, *hostRange, uint, error) {
+	if hl.hostCount == 0 {
+		return -1, nil, 0, ErrEmpty
+	}
+
+	if n < 0 {
+		return -1, nil, 0, errors.New("index can't be < 0")
+	}
+
+	if uint(n) >= hl.hostCount {
+		return -1, nil, 0, errors.New("index must be < hostCount")
+	}
+
+	var counted uint
+	for idx, hr := range hl.ranges {
+		rangeCount := hr.count()
+
+		if uint(n) <= rangeCount-1+counted {
+			return idx, hr, counted, nil
+		}
+		counted += rangeCount
+	}
+
+	// notreached?
+	return -1, nil, 0, errors.New("unknown error")
+}
+
+// Nth returns the string represenation of the n-th host in
+// the HostList. Returns an error if the index is invalid.
+func (hl *HostList) Nth(n int) (string, error) {
+	hl.RLock()
+	defer hl.RUnlock()
+
+	_, hr, depth, err := hl.getNthHostRange(n)
+	if err != nil {
+		return "", err
+	}
+
+	if hr.isRange {
+		return fmtRangeHost(hr, hr.lo+uint(n)-depth), nil
+	}
+	return hr.prefix, nil
+}
+
+func fmtRange(hr *hostRange) (rs string) {
+	rs = hr.prefix
+
+	if hr.isRange {
+		var lBr, rBr string
+		if hr.count() > 1 {
+			lBr = "["
+			rBr = "]"
+		}
+		rs = fmt.Sprintf("%s%s%s%s%s", hr.prefix, lBr, hr.rangedString(), rBr, hr.suffix)
+	}
+
+	return
+}
+
+// PopRange returns the string representation of the last
+// bracketed list of hosts. All hosts in the returned list
+// are removed from the HostList. Returns an error if the
+// HostList is empty.
+func (hl *HostList) PopRange() (hostRange string, err error) {
+	if hl.IsEmpty() {
```

Comment:
Similar question about taking the lock here.

Reply:
There are a few other places later in the file, but rather than clutter it with my questions, I'll leave it as a general question for all of them.